### PR TITLE
add stackoverflow basic service

### PR DESCRIPTION
**Pull request by lerushe**

### CODE REVIEW

File: backend/franky/service/stackoverflow/stackoverflow.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+import requests
+
+from model import UserData
+from service import Service
+
+
+class StackOverflow(Service):
+
+    def __init__(self):
+        self._rest_endpoint = 'https://api.stackexchange.com/2.2'
```
Comment (tcibinan) [inline backend/franky/service/stackoverflow/stackoverflow.py]: It looks good to me to add `rest_endpoint` constructor parameter with a default value `https://api.stackexchange.com/2.2`.

File: backend/franky/service/stackoverflow/stackoverflow.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+        self._service_name = 'stackoverflow'
+        self._user_url = 'users'
+        self._tags_url = 'tags'
+
+    def user(self, user_name):
```
Comment (tcibinan) [inline backend/franky/service/stackoverflow/stackoverflow.py]: Missing method return type parameter `UserData`.

File: backend/franky/service/stackoverflow/stackoverflow.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+        user = UserData(login=user_id, name=user_name, tags=user_tags)
+        return user
+
+    def _get_user_id(self, user_name) -> str:
+        payload = {'inname': user_name}
+        response_data = self._call(self._user_url, payload)
+        user_id = response_data['items'][0]['user_id']
+
+        if len(response_data['items']) > 1:
+            raise Exception('Username is not unique!')
```
Comment (tcibinan) [inline backend/franky/service/stackoverflow/stackoverflow.py]: Btw, I don't think we have to fail in the case. Probably, we should log a warning and return one of the user's data.

File: backend/franky/service/stackoverflow/stackoverflow.py [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    def user(self, user_name):
+        user_id = self._get_user_id(user_name)
+        user_tags = self._get_user_tags(user_id)
+        user = UserData(login=user_id, name=user_name, tags=user_tags)
+        return user
+
+    def _get_user_id(self, user_name) -> str:
+        payload = {'inname': user_name}
+        response_data = self._call(self._user_url, payload)
+        user_id = response_data['items'][0]['user_id']
```
Comment (tcibinan) [inline backend/franky/service/stackoverflow/stackoverflow.py]: As I see the code will fail with `ValueError` or ` IndexError` if a user is not found. Probably, we should handle the case and throw an exception with a proper error message. Some custom error type f.e. `StackoverflowServiceException` will be of help.

File: backend/franky/service/stackoverflow/stackoverflow.py [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+        payload = {'order': 'desc', 'sort': 'popular'}
+        request_url = f'{self._user_url}/{user_id}/{self._tags_url}'
+        response_data = self._call(request_url, payload)
+        tags = response_data['items']
+        tags_name = [tag['name'] for tag in tags]
+        return tags_name
+
+    def _call(self, request_url: str, payload: Dict) -> Dict:
+        url = urljoin(self._rest_endpoint, request_url)
+        payload['site'] = self._service_name
```
Comment (tcibinan) [inline backend/franky/service/stackoverflow/stackoverflow.py]: It is not a good practise to mutate function parameter values. The better approach will be to create a copy of a dictionary and then mutate it.

**Review by tcibinan — CHANGES_REQUESTED** [review]
Thanks for you work!

Please check a few comments below.

File: backend/franky/service/stackoverflow/stackoverflow.py [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+        return user
+
+    def _get_user_id(self, user_name: str) -> str:
+        payload = {'inname': user_name}
+        response_data = self._call(self._user_url, payload)
+        data = response_data['items']
+
+        if len(data) > 1:
+            user_id = self._get_unique_user(data, user_name)
+        elif len(data) == 1:
```
Comment (tcibinan) [inline backend/franky/service/stackoverflow/stackoverflow.py]: I suppose we should handle the `len(data) == 1` case the same way we handle `len(data) > 1`.

File: backend/franky/service/stackoverflow/stackoverflow.py [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+    def _get_user_tags(self, user_id: str) -> List:
+        payload = {'order': 'desc', 'sort': 'popular'}
+        request_url = f'{self._user_url}/{user_id}/{self._tags_url}'
+        response_data = self._call(request_url, payload)
+        tags_name = [tag['name'] for tag in response_data['items']]
+        return tags_name
+
+    def _call(self, request_url: str, payload: Dict) -> Dict:
+        url = urljoin(self._rest_endpoint, request_url)
+        full_payload = payload
```
Comment (tcibinan) [inline backend/franky/service/stackoverflow/stackoverflow.py]: It is not a copying of the dictionary. The `a = b` means that you copy not the object but its pointer to another variable. What I meant in the previous review was the following:
```
full_payload = dict(payload)
```
The operation above really *copies* the object rather then its pointer.

**Review by tcibinan — CHANGES_REQUESTED** [review]

**Review by tcibinan — APPROVED** [review]

**Review by dyadyaJora — APPROVED** [review]

**Pull request merged**